build/templates/matchers.py.mako

```diff
@@ -89,23 +89,53 @@ class ViStringMatcher(object):
 # Custom Type
 
 
+def _compare_ctype_structs(expected, actual):
+    # From https://stackoverflow.com/questions/20986330/print-all-fields-of-ctypes-structure-with-introspection
+    for field in expected._fields_:
+        field_name = field[0]
+        expected_val = getattr(expected, field_name)
+        actual_val = getattr(actual, field_name)
+        if expected_val != actual_val:
+            print("Unexpected value field {0}. Expected: {1}. Received: {2}".format(field_name, expected_val, actual_val))
+            return False
+    return True
+
+
 class CustomTypeMatcher(object):
     def __init__(self, expected_type, expected_value):
         self.expected_type = expected_type
         self.expected_value = expected_value
 
-    def __eq__(self, other):
-        if not isinstance(other, self.expected_type):
-            print("Unexpected type. Expected: {0}. Received: {1}".format(self.expected_type, type(other)))
+    def __eq__(self, actual):
+        if not isinstance(actual, self.expected_type):
+            print("Unexpected type. Expected: {0}. Received: {1}".format(self.expected_type, type(actual)))
             return False
-        # From https://stackoverflow.com/questions/20986330/print-all-fields-of-ctypes-structure-with-introspection
-        for field in self.expected_value._fields_:
-            field_name = field[0]
-            expected_val = getattr(self.expected_value, field_name)
-            other_val = getattr(other, field_name)
-            if expected_val != other_val:
-                print("Unexpected value field {0}. Expected: {1}. Received: {2}".format(field_name, expected_val, other_val))
-                return False
+        return _compare_ctype_structs(self.expected_value, actual)
+
+
+class CustomTypeBufferMatcher(object):
+    def __init__(self, expected_element_type, expected_value):
+        self.expected_value = expected_value
+        self.expected_size = len(expected_value)
+        self.expected_type = expected_element_type * self.expected_size
+        self.expected_element_type = expected_element_type
+
+    def __eq__(self, actual):
+        if not isinstance(actual, self.expected_type):
+            print("Unexpected array type. Expected: {0}. Received: {1}".format(self.expected_type, type(actual)))
+            return False
+        if self.expected_size != len(actual):
+            print("Unexpected length. Expected: {0}. Received: {1}".format(self.expected_size, len(actual)))
+            return False
+        if self.expected_value is not None:
+            # Can't compare the objects directly because they're different types (one is list, another is ctypes array).
+            # Go element by element, which allows for reporting the first index where different values were found.
+            for i in range(0, len(self.expected_value)):
```

Comment:
I think the Pythonic way of doing this is to use zip. You can give it a shot. Something like:

```
for i, j in zip(self.expected_value , actual):
```

(we've been doing it wrong all over the place, too much ancient C++ under our belts)

Reply:
... This only works after you verified len() is the same:

```
>>> zip([1,2], [1,2,3])
[(1, 1), (2, 2)]
```

Reply:
`zip()` works, but it has memory implications. Because it copies each of the lists, if they are large, then you've just doubled the memory required.

It may be that here it is okay since we will never test with very large lists/arrays.

Reply:
I don't think so. At least not in Python 3.

> Make an iterator that aggregates elements from each of the iterables.

https://docs.python.org/3/library/functions.html#zip

Reply:
You're right about Python 3, but we support both Python 2 as well.

Reply:
IMHO, there's no need to support the _code generator_ on Python 3. Just the generated code.

Reply:
Do we then drop Python 2 from Travis? It first builds then tests, which means the code generator has to support Python 2.

Reply:
I think it's tricky to do because of how Travis works.

Ideally we would

1. Run unit tests for build in the one version of Python that we support.
2. Build the code in the one version of Python that we support.
3. Run unit tests on all versions of Python that we support.

But given that each version of Python is in its own Docker container and that we must unit test generated code on each version of Python, we have no choice but to build in each version of Python unless we do some major restructuring of the project.

So because of this, it seems simpler to keep ensuring that the generator runs on all supported version of Python for the foreseeable future. But I don't care if zip() uses more RAM in the Python 2 builds :)